File: luci/applications/luci-app-acctl/luasrc/controller/acctl.lua
--[[
LuCI - Lua Configuration Interface
AC Controller — LuCI Controller Module
JSON backend (replaces SQLite)

Licensed under the Apache License, Version 2.0
http://www.apache.org/licenses/LICENSE-2.0
]]--

module("luci.controller.acctl", package.seeall)

local sys   = require "luci.sys"
local util  = require "luci.util"
local http  = require "luci.http"
local uci   = require "luci.model.uci".cursor()

-- ========================================================================
-- Helper: run acctl-cli and parse JSON response
-- ========================================================================

local function cli_output(cmd)
    local f = io.popen(cmd .. " 2>/dev/null")
    if not f then return nil end
    local r = f:read("*a")
    f:close()
    return r
end

local function cli_json(args)
    local cmd = "acctl-cli " .. args
    local out = cli_output(cmd)
    if not out or out == "" then return nil end
    local ok, data = pcall(http.parse_json, out)
    if not ok then return nil end
    return data
end

local function is_running()
    return sys.call("pgrep -x acser > /dev/null 2>&1") == 0
end

-- ========================================================================
-- Menu entries
-- ========================================================================

function index()
    entry({"admin", "network", "acctl"},
        alias("admin", "network", "acctl", "general"),
        _("AC Controller"), 60).dependent = false

    entry({"admin", "network", "acctl", "general"},
        cbi("acctl/general", {autoapply=true}),
        _("General"), 10)

    entry({"admin", "network", "acctl", "ap_list"},
        cbi("acctl/ap_list", {autoapply=true}),
        _("AP List"), 20)

    entry({"admin", "network", "acctl", "groups"},
        cbi("acctl/groups", {autoapply=true}),
        _("AP Groups"), 25)

    entry({"admin", "network", "acctl", "templates"},
        cbi("acctl/templates", {autoapply=true}),
        _("Templates"), 30)

    entry({"admin", "network", "acctl", "alarms"},
        cbi("acctl/alarms", {autoapply=true}),
        _("Alarms"), 40)

    entry({"admin", "network", "acctl", "firmware"},
        cbi("acctl/firmware", {autoapply=true}),
        _("Firmware"), 50)

    entry({"admin", "network", "acctl", "system"},
        cbi("acctl/system"),
        _("System"), 60)

    -- REST API endpoints
    entry({"admin", "network", "acctl", "api", "status"},
        call("api_status"))

    entry({"admin", "network", "acctl", "api", "aps"},
        call("api_aps"))

    entry({"admin", "network", "acctl", "api", "aps_action"},
        call("api_aps_action"))

    entry({"admin", "network", "acctl", "api", "alarms"},
        call("api_alarms"))

    entry({"admin", "network", "acctl", "api", "alarms_ack"},
        call("api_alarms_ack"))

    entry({"admin", "network", "acctl", "api", "groups"},
        call("api_groups"))

    entry({"admin", "network", "acctl", "api", "firmwares"},
        call("api_firmwares"))

    entry({"admin", "network", "acctl", "api", "cmd"},
        call("api_cmd"))

    entry({"admin", "network", "acctl", "api", "restart"},
        call("api_restart"))
end

-- ========================================================================
-- API: Status
-- ========================================================================

function api_status()
    local stats = cli_json("stats") or {}
    local status = {
        running       = is_running(),
        ap_online     = tonumber(stats.ap_online) or 0,
        ap_total      = tonumber(stats.ap_total)  or 0,
        alarm_count   = tonumber(stats.alarms)    or 0,
        pool_left     = 0,
        pool_total    = 0,
        timestamp     = os.time()
    }

    http.prepare_content("application/json")
    http.write_json(status)
end

-- ========================================================================
-- API: AP List
-- ========================================================================

function api_aps()
    local mac    = http.formvalue("mac")
    local result = {}

    if mac and mac ~= "" then
        -- Single AP detail — load all and find by mac
        local data = cli_json("aps --limit 500") or {}
        if data and data.aps then
            for _, ap in ipairs(data.aps) do
                if ap.mac == mac then
                    result.mac          = ap.mac or ""
                    result.hostname     = ap.hostname or ""
                    result.wan_ip       = ap.wan_ip or ""
                    result.wifi_ssid    = ap.wifi_ssid or ""
                    result.firmware     = ap.firmware or ""
                    result.online_users = tonumber(ap.online_users) or 0
                    result.online       = (tonumber(ap.device_down) == 0)
                    result.last_seen    = tonumber(ap.last_seen) or 0
                    result.group_id     = tonumber(ap.group_id) or 0
                    break
                end
            end
        end
    else
        -- All APs
        result.aps   = {}
        result.count = 0
        local data = cli_json("aps --limit 500") or {}
        if data and data.aps then
            for _, ap in ipairs(data.aps) do
                result.count = result.count + 1
                result.aps[#result.aps + 1] = {
                    mac          = ap.mac or "",
                    hostname     = ap.hostname or "",
                    wan_ip       = ap.wan_ip or "",
                    wifi_ssid    = ap.wifi_ssid or "",
                    firmware     = ap.firmware or "",
                    online_users = tonumber(ap.online_users) or 0,
                    online       = (tonumber(ap.device_down) == 0),
                    last_seen    = tonumber(ap.last_seen) or 0,
                    group_id     = tonumber(ap.group_id) or 0
                }
            end
        end
    end

    http.prepare_content("application/json")
    http.write_json(result)
end

-- ========================================================================
-- API: AP Bulk Actions
-- ========================================================================

function api_aps_action()
    local action = http.formvalue("action")
    local macs   = http.formvalue("macs") or ""
    local result = { code = 0, message = "", affected = 0 }

    if action == "" or macs == "" then
        result.code    = 400
        result.message = "action and macs required"
        http.prepare_content("application/json")
        http.write_json(result)
        return
    end

    -- Validate action against whitelist
    local valid_actions = { reboot = true, config = true, upgrade = true }
    if not valid_actions[action] then
        result.code    = 400
        result.message = "Invalid action"
        http.prepare_content("application/json")
        http.write_json(result)
        return
    end

    -- Audit log (use only validated action, sanitize macs for shell safety)
    local safe_macs = macs:gsub("[^%x:, ]", "")
    cli_output(string.format(
        "acctl-cli audit admin %s ap_batch '%s' '' '' ''",
        action, safe_macs:sub(1, 50)))

    if action == "reboot" then
        result.message = "Reboot command queued for APs"
    elseif action == "config" then
        result.message = "Configuration push queued"
    elseif action == "upgrade" then
        result.message = "Firmware upgrade queued"
    else
        result.code    = 400
        result.message = "Unknown action"
    end

    http.prepare_content("application/json")
    http.write_json(result)
end

-- ========================================================================
-- API: Alarm List
-- ========================================================================

function api_alarms()
    local limit = tonumber(http.formvalue("limit")) or 50
    local data = cli_json("alarms --limit " .. tostring(limit)) or {}

    local alarms = {}
    if data.alarms then
        for _, a in ipairs(data.alarms) do
            local level = 0
            if a.level == "warn"      then level = 1
            elseif a.level == "error" then level = 2
            elseif a.level == "critical" then level = 3
            end
            alarms[#alarms + 1] = {
                id            = tonumber(a.id) or 0,
                ap_mac        = a.mac or "",
                level         = a.level or "info",
                level_num     = level,
                message       = a.message or "",
                acknowledged  = (a.ack == 1 or a.ack == true),
                created_at    = a.ts or ""
            }
        end
    end

    http.prepare_content("application/json")
    http.write_json({alarms = alarms, count = #alarms})
end

-- ========================================================================
-- API: Acknowledge Alarms
-- ========================================================================

function api_alarms_ack()
    local ids = http.formvalue("ids") or ""
    local result = { code = 0, acknowledged = 0 }

    if ids == "all" then
        cli_output("acctl-cli ack-all")
        result.acknowledged = -1
    else
        for id in ids:gmatch("%d+") do
            cli_output("acctl-cli ack " .. id)
            result.acknowledged = result.acknowledged + 1
        end
    end

    http.prepare_content("application/json")
    http.write_json(result)
end

-- ========================================================================
-- API: AP Groups
-- ========================================================================

function api_groups()
    local data = cli_json("groups") or {}
    local groups = {}

    if data.groups then
        -- Count APs per group from the full AP list
        local aps_data = cli_json("aps --limit 500") or {}
        local ap_count = {}
        if aps_data and aps_data.aps then
            for _, ap in ipairs(aps_data.aps) do
                local gid = tonumber(ap.group_id) or 0
                ap_count[gid] = (ap_count[gid] or 0) + 1
            end
        end

        for _, grp in ipairs(data.groups) do
            groups[#groups + 1] = {
                id          = tonumber(grp.id) or 0,
                name        = grp.name or "",
                description = grp.description or "",
                policy      = grp.policy or "manual",
                ap_count    = ap_count[tonumber(grp.id) or 0] or 0
            }
        end
    end

    http.prepare_content("application/json")
    http.write_json({groups = groups})
end

-- ========================================================================
-- API: Firmware List
-- ========================================================================

function api_firmwares()
    local data = cli_json("firmware") or {}

    local firmwares = {}
    if data.firmwares then
        for _, fw in ipairs(data.firmwares) do
            firmwares[#firmwares + 1] = {
                version     = fw.version or "",
                filename    = fw.filename or "",
                size        = tonumber(fw.size) or 0,
                sha256      = fw.sha256 or "",
                uploaded_at = fw.uploaded_at or ""
            }
        end
    end

    http.prepare_content("application/json")
    http.write_json({firmwares = firmwares})
end

-- ========================================================================
-- API: Command Execution
-- ========================================================================

function api_cmd()
    local mac = http.formvalue("mac")
    local cmd = http.formvalue("cmd")
    local result = { code = 0, message = "" }

    if not mac or not cmd or mac == "" or cmd == "" then
        result.code    = 400
        result.message = "mac and cmd are required"
        http.prepare_content("application/json")
        http.write_json(result)
        return
    end

    -- Command whitelist validation
    local allowed = {
        ["reboot"]              = true,
        ["uptime"]              = true,
        ["ifconfig"]            = true,
        ["iwconfig"]            = true,
        ["wifi"]                = true,
        ["cat /proc/uptime"]    = true,
        ["cat /proc/loadavg"]   = true,
        ["cat /tmp/ap_status"]  = true,
    }

    -- Exact match: command must exactly equal whitelist entry
    -- Reject if cmd contains any shell metacharacters or control chars
    local dangerous_patterns = {
        ";",     -- command chaining
        "|",     -- pipe
        "`",     -- command substitution
        "$(" ,   -- command substitution variant
        "&",     -- background
        "&&",    -- AND chaining
        "||",    -- OR chaining
        ">",     -- output redirect
        "<",     -- input redirect
        ">>",    -- append redirect
        "<<",    -- heredoc
        "~/",    -- home dir expansion
        "/etc/", -- system config access
        "/bin/", -- binary dir access
        "/usr/", -- usr dir access
        "/var/", -- var dir access
        "/dev/", -- device access
        "/proc/",-- procfs access
        "/root/",-- root dir access
        "../",   -- directory traversal
        "0>",    -- fd redirect
        "1>",    -- stdout redirect
        "2>",    -- stderr redirect
        "nohup", -- background process
        "screen",-- screen sessions
        "tmux",  -- tmux sessions
        "wget",  -- network download
        "curl",  -- network download
        "nc ",   -- netcat
        "ncat",  -- netcat variant
        "python",-- python interpreter
        "perl",  -- perl interpreter
        "ruby",  -- ruby interpreter
        "php",   -- php interpreter
        "bash",  -- bash shell
        "sh -c", -- explicit shell
        "exec",  -- exec call
        "eval",  -- eval
        "chmod", -- permission change
        "chown", -- ownership change
        "rm -rf",-- recursive delete
        "mkfs",  -- filesystem creation
        "dd ",   -- raw disk write
        "fdisk", -- disk partitioning
        "mount", -- mount filesystem
        "umount",-- unmount filesystem
        "passwd",-- password change
        "su ",   -- switch user
        "sudo",  -- privilege escalation
        "shutdown", -- system shutdown
        "halt",  -- halt system
        "poweroff", -- power off
    }

    -- Check for dangerous patterns
    for _, pattern in ipairs(dangerous_patterns) do
        if cmd:find(pattern, 1, true) then
            result.code    = 403
            result.message = "Command contains forbidden characters"
            http.prepare_content("application/json")
            http.write_json(result)
            return
        end
    end

    -- Reject if command contains path separators not in whitelist
    if cmd:match("/") and not allowed[cmd] then
        result.code    = 403
        result.message = "Command contains forbidden characters"
        http.prepare_content("application/json")
        http.write_json(result)
        return
    end

    -- Exact whitelist match
    local found = allowed[cmd] == true

    if not found then
        result.code    = 403
        result.message = "Command not in whitelist"
        http.prepare_content("application/json")
        http.write_json(result)
        return
    end

    -- Audit (sanitize inputs for shell safety)
    local safe_mac = mac:gsub("[^%x:]", ""):sub(1, 20)
    local safe_cmd = cmd:gsub("'", "'\\''"):sub(1, 100)
    cli_output(string.format(
        "acctl-cli audit admin EXEC ap '%s' '' '%s' ''",
        safe_mac, safe_cmd))

    result.message = "Command queued successfully"
    http.prepare_content("application/json")
    http.write_json(result)
end

-- ========================================================================
-- API: Restart Service
-- ========================================================================

function api_restart()
    local result = { code = 0, message = "" }
    if is_running() then
        sys.exec("/etc/init.d/acctl restart > /dev/null 2>&1")
        result.message = "AC Controller restarting..."
    else
        sys.exec("/etc/init.d/acctl start > /dev/null 2>&1")
        result.message = "AC Controller starting..."
    end
    http.prepare_content("application/json")
    http.write_json(result)
end

File: luci24/usr/share/ucode/luci/controller/acctl.js
// LuCI - AC Controller JavaScript Controller
// Compatible with OpenWrt 24.x (ucode-based LuCI)

'use strict';

import { curent } from 'uci';
import { exec, call } from 'ubus';

return class ACctlController extends LuCI.controller {
    constructor() {
        super();
        this.acctl_cli = '/usr/bin/acctl-cli';
    }

    getCLI(cmd) {
        try {
            const f = exec(this.acctl_cli + ' ' + cmd + ' 2>/dev/null');
            if (!f || f.stdout === '') return null;
            return JSON.parse(f.stdout);
        } catch (e) {
            return null;
        }
    }

    isRunning() {
        try {
            exec('pgrep -x acser > /dev/null 2>&1');
            return call('service', 'check', { name: 'acctl' }) === 0;
        } catch (e) {
            return false;
        }
    }

    registerI18n() {
        return _('AC Controller');
    }

    index() {
        super.entry({
            title: _('AC Controller'),
            path: '/admin/network/acctl',
            icon: 'signal',
        }, super.chain(
            super.cart('admin/network/acctl/general'),
            super.cart('admin/network/acctl/ap_list'),
            super.cart('admin/network/acctl/groups'),
            super.cart('admin/network/acctl/alarms'),
            super.cart('admin/network/acctl/firmware'),
            super.cart('admin/network/acctl/system')
        ));

        super.entry({
            title: _('AC Controller Status'),
            path: '/admin/services/acctl',
            icon: 'signal',
        }, super.cart('admin/services/acctl/status'));
    }

    apiStatus() {
        const stats = this.getCLI('stats') || {};
        const running = this.isRunning();
        const response = {
            running: running,
            ap_online: parseInt(stats.ap_online) || 0,
            ap_total: parseInt(stats.ap_total) || 0,
            alarm_count: parseInt(stats.alarms) || 0,
            pool_left: 0,
            pool_total: 0,
            timestamp: Math.floor(Date.now() / 1000)
        };
        http.write_json(response);
    }

    apiAps() {
        const mac = http.formvalue('mac');
        const result = { aps: [], count: 0 };

        if (mac && mac !== '') {
            const data = this.getCLI('aps --limit 500');
            if (data && data.aps) {
                for (const ap of data.aps) {
                    if (ap.mac === mac) {
                        result.mac = ap.mac || '';
                        result.hostname = ap.hostname || '';
                        result.wan_ip = ap.wan_ip || '';
                        result.wifi_ssid = ap.wifi_ssid || '';
                        result.firmware = ap.firmware || '';
                        result.online_users = parseInt(ap.online_users) || 0;
                        result.online = parseInt(ap.device_down) === 0;
                        result.last_seen = parseInt(ap.last_seen) || 0;
                        result.group_id = parseInt(ap.group_id) || 0;
                        break;
                    }
                }
            }
        } else {
            const data = this.getCLI('aps --limit 500');
            if (data && data.aps) {
                for (const ap of data.aps) {
                    result.aps.push({
                        mac: ap.mac || '',
                        hostname: ap.hostname || '',
                        wan_ip: ap.wan_ip || '',
                        wifi_ssid: ap.wifi_ssid || '',
                        firmware: ap.firmware || '',
                        online_users: parseInt(ap.online_users) || 0,
                        online: parseInt(ap.device_down) === 0,
                        last_seen: parseInt(ap.last_seen) || 0,
                        group_id: parseInt(ap.group_id) || 0
                    });
                    result.count++;
                }
            }
        }
        http.write_json(result);
    }

    apiApsAction() {
        const action = http.formvalue('action');
        const macs = http.formvalue('macs') || '';
        const result = { code: 0, message: '', affected: 0 };

        if (action === '' || macs === '') {
            result.code = 400;
            result.message = 'action and macs required';
            http.write_json(result);
            return;
        }

        const valid_actions = { reboot: true, config: true, upgrade: true };
        if (!valid_actions[action]) {
            result.code = 400;
            result.message = 'Invalid action';
            http.write_json(result);
            return;
        }

        const safe_macs = macs.replace(/[^%x:, ]/g, '');
        this.getCLI(`audit admin ${action} ap_batch '${safe_macs.substring(0, 50)}' '' '' ''`);

        if (action === 'reboot') result.message = 'Reboot command queued for APs';
        else if (action === 'config') result.message = 'Configuration push queued';
        else if (action === 'upgrade') result.message = 'Firmware upgrade queued';

        http.write_json(result);
    }

    apiAlarms() {
        const limit = parseInt(http.formvalue('limit')) || 50;
        const data = this.getCLI('alarms --limit ' + limit) || { alarms: [] };
        const alarms = [];

        for (const a of (data.alarms || [])) {
            let level = 0;
            if (a.level === 'warn') level = 1;
            else if (a.level === 'error') level = 2;
            else if (a.level === 'critical') level = 3;

            alarms.push({
                id: parseInt(a.id) || 0,
                ap_mac: a.mac || '',
                level: a.level || 'info',
                level_num: level,
                message: a.message || '',
                acknowledged: a.ack == 1 || a.ack === true,
                created_at: a.ts || ''
            });
        }
        http.write_json({ alarms: alarms, count: alarms.length });
    }

    apiAlarmsAck() {
        const ids = http.formvalue('ids') || '';
        const result = { code: 0, acknowledged: 0 };

        if (ids === 'all') {
            this.getCLI('ack-all');
            result.acknowledged = -1;
        } else {
            const idList = ids.match(/\d+/g) || [];
            for (const id of idList) {
                this.getCLI('ack ' + id);
                result.acknowledged++;
            }
        }
        http.write_json(result);
    }

    apiGroups() {
        const data = this.getCLI('groups') || { groups: [] };
        const groups = [];

        const aps_data = this.getCLI('aps --limit 500') || { aps: [] };
        const ap_count = {};
        for (const ap of (aps_data.aps || [])) {
            const gid = parseInt(ap.group_id) || 0;
            ap_count[gid] = (ap_count[gid] || 0) + 1;
        }

        for (const grp of (data.groups || [])) {
            groups.push({
                id: parseInt(grp.id) || 0,
                name: grp.name || '',
                description: grp.description || '',
                policy: grp.policy || 'manual',
                ap_count: ap_count[parseInt(grp.id) || 0] || 0
            });
        }
        http.write_json({ groups: groups });
    }

    apiFirmwares() {
        const data = this.getCLI('firmware') || { firmwares: [] };
        const firmwares = [];

        for (const fw of (data.firmwares || [])) {
            firmwares.push({
                version: fw.version || '',
                filename: fw.filename || '',
                size: parseInt(fw.size) || 0,
                sha256: fw.sha256 || '',
                uploaded_at: fw.uploaded_at || ''
            });
        }
        http.write_json({ firmwares: firmwares });
    }

    apiCmd() {
        const mac = http.formvalue('mac');
        const cmd = http.formvalue('cmd');
        const result = { code: 0, message: '' };

        if (!mac || !cmd || mac === '' || cmd === '') {
            result.code = 400;
            result.message = 'mac and cmd are required';
            http.write_json(result);
            return;
        }

        const allowed = {
            'reboot': true,
            'uptime': true,
            'ifconfig': true,
            'iwconfig': true,
            'wifi': true,
            'cat /proc/uptime': true,
            'cat /proc/loadavg': true,
            'cat /tmp/ap_status': true
        };

        const dangerous = [';', '|', '`', '$(', '&', '&&', '||', '>', '<', '>>',
            '/bin/', '/etc/', '/usr/', '/var/', '/dev/', '/proc/', '/root/', '../',
            'nohup', 'screen', 'tmux', 'wget', 'curl', 'python', 'perl', 'ruby',
            'php', 'bash', 'sh -c', 'exec', 'eval', 'chmod', 'chown', 'rm -rf',
            'mkfs', 'dd', 'fdisk', 'mount', 'umount', 'passwd', 'su', 'sudo'];

        for (const p of dangerous) {
            if (cmd.indexOf(p) !== -1) {
                result.code = 403;
                result.message = 'Command contains forbidden patterns';
                http.write_json(result);
                return;
            }
        }

        if (!allowed[cmd]) {
            result.code = 403;
            result.message = 'Command not in whitelist';
            http.write_json(result);
            return;
        }

        const safe_mac = mac.replace(/[^%x:]/g, '').substring(0, 20);
        this.getCLI(`audit admin EXEC ap '${safe_mac}' '' '${cmd}' ''`);

        result.message = 'Command queued successfully';
        http.write_json(result);
    }

    apiRestart() {
        const result = { code: 0, message: '' };
        try {
            exec('/etc/init.d/acctl restart > /dev/null 2>&1');
            result.message = this.isRunning() ? 'AC Controller restarting...' : 'AC Controller starting...';
        } catch (e) {
            result.code = 500;
            result.message = 'Failed to restart service';
        }
        http.write_json(result);
    }

    apiAction() {
        const action = http.formvalue('action');
        const result = { code: 0, message: '' };

        try {
            if (action === 'start') {
                exec('/etc/init.d/acctl start > /dev/null 2>&1');
                result.message = 'AC Controller started';
            } else if (action === 'stop') {
                exec('/etc/init.d/acctl stop > /dev/null 2>&1');
                result.message = 'AC Controller stopped';
            } else if (action === 'restart') {
                exec('/etc/init.d/acctl restart > /dev/null 2>&1');
                result.message = 'AC Controller restarted';
            } else {
                result.code = 400;
                result.message = 'Invalid action';
            }
        } catch (e) {
            result.code = 500;
            result.message = 'Failed to execute action';
        }
        http.write_json(result);
    }

    apiSwitchMode() {
        const mode = http.formvalue('mode');
        const result = { code: 0, message: '' };

        try {
            if (mode === 'ac' || mode === 'ap') {
                const uci = curent();
                uci.set('acctl', 'acctl', 'mode', mode);
                uci.commit('acctl');
                exec('/etc/init.d/acctl restart > /dev/null 2>&1');
                result.message = 'Mode switched to ' + (mode === 'ac' ? 'AC' : 'AP') + ' mode';
            } else {
                result.code = 400;
                result.message = 'Invalid mode';
            }
        } catch (e) {
            result.code = 500;
            result.message = 'Failed to switch mode';
        }
        http.write_json(result);
    }
});